nx/lib/nx/defn/expr.ex

```diff
@@ -604,6 +604,94 @@ defmodule Nx.Defn.Expr do
     end
   end
 
+  defp while_range(
+         range,
+         file,
+         line,
+         initial,
+         %T{vectorized_axes: [{first_axis, _} | _] = vectorized_axes} = generator,
+         condition_body,
+         opts
+       ) do
+    num_vectorized_axes = length(vectorized_axes)
+
+    {initial, max_vec_size} =
+      Composite.traverse(initial, nil, fn leaf, _ ->
+        # broadcast and pull common axes to the front
+        [_, leaf] = Nx.broadcast_vectors([generator, leaf])
+
+        %{vectorized_axes: leaf_axes} = leaf
+
+        # remove common axes so that we can flatten them into a fixed axis.
+        # We use the `first_axis` to ensure that there's no name collision
+        # in an easy way.
+        vectorized_axes = [{first_axis, :auto} | Enum.drop(leaf_axes, num_vectorized_axes)]
+
+        %{vectorized_axes: [{_, s} | _]} = leaf = Nx.revectorize(leaf, vectorized_axes)
+        {leaf, s}
+      end)
+
+    flat_generator = generator |> Nx.revectorize([{first_axis, :auto}]) |> Nx.devectorize()
+    outer_initial = {{tensor(0), flat_generator}, initial}
+
+    {{{index_param, inner_gen_param}, outer_arg}, outer_context} =
+      to_param_expr(outer_initial, :while)
+
+    outer_condition = Nx.less(index_param, tensor(max_vec_size))
+
+    inner_initial =
+      Composite.traverse(outer_arg, fn %{vectorized_axes: [_ | ax]} = leaf ->
+        leaf
+        |> Nx.devectorize()
+        |> Nx.take(index_param)
+        |> Nx.vectorize(ax)
+      end)
+
+    inner_while =
+      while_range(
+        range,
+        file,
+        line,
+        inner_initial,
+        Nx.take(inner_gen_param, index_param),
+        condition_body,
+        opts
+      )
+
+    {outer_body, _} =
+      Composite.traverse(
+        inner_while,
+        Composite.flatten_list([outer_arg]),
+        fn node, [target | targets] ->
+          target = Nx.devectorize(target)
+          starts = [index_param | List.duplicate(0, tuple_size(target.shape) - 1)]
+
+          updated_target =
+            target
+            |> Nx.put_slice(starts, Nx.devectorize(node) |> Nx.new_axis(0))
+            |> Nx.vectorize(target.vectorized_axes)
+
+          {updated_target, targets}
+        end
+      )
```

Comment:
Do you think it is worth sharing some of this logic with the other `while` branch? I think many of these `traverse` calls are identical.

Reply:
Yes. The code is mostly copy-paste :)
I just didn't do it before because I wanted to validate the general idea and behavior of the code with you first

Reply:
Yeah, looks good to me. :) I think we can probably share all of its contents and make the "inner_while" bit an argument?

Reply:
I shared a prefix section and a suffix section of the code. the building of the inner_while seemed different enough to me to warrant keeping it in place instead of passing that section as an anonymous function or whatever